src/charm.py

```diff
@@ -31,24 +32,45 @@ def __init__(self, *args: typing.Any):
         """
         super().__init__(*args)
         self._haproxy_route = HaproxyRouteRequirer(self, HAPROXY_ROUTE_RELATION)
+        self._ingress = IngressPerAppProvider(self, INGRESS_RELATION)
         self.framework.observe(self.on.config_changed, self._reconcile)
         self.framework.observe(self.on[HAPROXY_ROUTE_RELATION].relation_changed, self._reconcile)
         self.framework.observe(self.on[HAPROXY_ROUTE_RELATION].relation_broken, self._reconcile)
         self.framework.observe(self.on[HAPROXY_ROUTE_RELATION].relation_departed, self._reconcile)
+        self.framework.observe(self._ingress.on.data_provided, self._reconcile)
+        self.framework.observe(self._ingress.on.data_removed, self._reconcile)
 
-    @validate_config
     def _reconcile(self, _: ops.EventBase) -> None:
         """Refresh haproxy-route requirer data."""
-        integrator_information = IntegratorInformation.from_charm(self)
-        if not self._haproxy_route.relation:
-            self.unit.status = ops.BlockedStatus("haproxy-route relation missing.")
-            return
-        self._haproxy_route.provide_haproxy_route_requirements(
-            service=f"{self.model.name}-{self.app.name}",
-            ports=[integrator_information.backend_port],
-            unit_address=str(integrator_information.backend_address),
-        )
-        self.unit.status = ops.ActiveStatus()
+        try:
+            if not self._haproxy_route.relation:
+                self.unit.status = ops.BlockedStatus("Missing haproxy-route relation.")
+                return
+            mode = configurator.get_mode(
+                self, self.model.get_relation(self._ingress.relation_name)
+            )
+            if mode == configurator.Mode.INTEGRATOR:
+                integrator_information = configurator.IntegratorInformation.from_charm(self)
+                self._haproxy_route.provide_haproxy_route_requirements(
+                    service=f"{self.model.name}-{self.app.name}",
+                    ports=[integrator_information.backend_port],
+                    unit_address=str(integrator_information.backend_address),
+                )
+            elif mode == configurator.Mode.ADAPTER:
+                relation = self.model.get_relation(self._ingress.relation_name)
+                data = self._ingress.get_data(relation)
+                self._haproxy_route.provide_haproxy_route_requirements(
+                    service=f"{data.app.model}-{data.app.name}",
+                    ports=[data.app.port],
+                    unit_address=str([udata.host for udata in data.units][0]),
```

Comment:
Eventually forward the first value of the 'proxied_urls' of the haproxy-route provider to the ingress requirer